### PR TITLE
Start of work to add spack.

### PR DESCRIPTION
**Pull request by vsoch**
Hi repology! This is the start of work to add the spack package manager, discussed in https://github.com/spack/spack/issues/25288.

I had a really hard time reproducing the production environment (installing all those dependencies and custom extension libraries on my host is no small thing!) so I had two options - either looking into updating the current setup to allow a simpler database like sqlite3, or providing a full development environment with containers. Since I would suspect there are reasons for wanting postgres, I opted for the latter, and have added a docker-compose.yml file for running two services, a database and application container (details in the README).

I think I have most of it working, but I'm missing some detail about how to add metadata for the packages. Specifically I see:

```
root@91bf370b9026:/code# ./repology-update.py --fetch --fetch --parse --database --postupdate --dsn "host=db dbname=repology user=repology password=repology" --enabled-repositories=spack
Aug 11 04:23:04 loading rules
/usr/local/lib/python3.9/site-packages/rpm.py:15: UserWarning: The RPM Python bindings are not currently available via PyPI.

Please install them with your distro package manager (typically called
'python2-rpm' or 'python3-rpm'), and ensure that any virtual environments
needing the API are configured to be able to see the system site packages
directory.

  warnings.warn(warning_msg)
Aug 11 04:23:06 updating repositories metadata
Aug 11 04:23:06   committing changes
Aug 11 04:23:06 not fetching spack to honor update period (0:07:49.867128 left)
Aug 11 04:23:15 not parsing spack due to no data changes since last run
Aug 11 04:23:15 starting the update
Aug 11 04:23:15 forcing update for projects affected by changed CPEs
Aug 11 04:23:15 updating projects
Aug 11 04:23:15 WARNING: parsed packages for repository Spack are missing, treating repository as empty
Aug 11 04:23:15 ERROR: no parsed packages found
Aug 11 04:23:15   done: added 0, removed 0, changed 0, unchanged 0, total change 0.00%
Aug 11 04:23:15 updating field statistics
Aug 11 04:23:15 explicit analyze is disabled
Aug 11 04:23:15 creating new links
Aug 11 04:23:15 translating incoming packages
Aug 11 04:23:15 preparing updated packages
Aug 11 04:23:15 updating CPE information
Aug 11 04:23:15 updating vulnerabilities
Aug 11 04:23:15 updating projects
Aug 11 04:23:15 updating maintainers (precreate)
Aug 11 04:23:15 updating maintainers
Aug 11 04:23:15 updating repositories
Aug 11 04:23:15 updating tracks
Aug 11 04:23:15 updating track versions
Aug 11 04:23:15 updating project releases
Aug 11 04:23:15 updating project events
Aug 11 04:23:15 updating maintainer events
Aug 11 04:23:15 updating repository events
Aug 11 04:23:15 updating projects turnover
Aug 11 04:23:15 updating links
Aug 11 04:23:15 updating statistics (delta)
Aug 11 04:23:15 updating redirects
Aug 11 04:23:15 applying updated packages
Aug 11 04:23:15 updating binding table repo_metapackages
Aug 11 04:23:15 updating binding table category_metapackages
Aug 11 04:23:15 updating binding table maintainer_metapackages
Aug 11 04:23:15 updating binding table maintainer_and_repo_metapackages
Aug 11 04:23:15 updating project names
Aug 11 04:23:15 updating url relations (all)
Aug 11 04:23:15 updating url relations (filtered)
Aug 11 04:23:15 updating projects has_related flag
Aug 11 04:23:15 updating problems
Aug 11 04:23:15 updating repository problem counts
Aug 11 04:23:15 updating repository maintainers
Aug 11 04:23:15 updating repository maintainer counts
Aug 11 04:23:15 updating statistics (global)
Aug 11 04:23:15 updating histories
Aug 11 04:23:15 finalizing the update
Aug 11 04:23:15 committing changes
Aug 11 04:23:15 performing database post-update actions
Aug 11 04:23:15   committing changes
Aug 11 04:23:15 total time taken: 11.23 seconds
```
It seems that even though I am iterating through packages using the factory/pkg and adding metadata, there are none found, so I must be missing something. Could you take a look and please advise? Thank you!

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

### CODE REVIEW

**Review by AMDmi3 — CHANGES_REQUESTED** [review]
Please remove everything related to docker, it won't be accepted. For the note, it should've been in a separate pull request.

**Comment by AMDmi3** [comment]
From a quick glance, I don't see why no packages are parsed. You can check parsing logs in the web UI for your instance (Repositories → Updates). The code looks mostly good, I can finish this up myself as soon as it's cleared of unrelated changes.

**Comment by vsoch** [comment]
Sure happy to do another PR with the Docker work! I need to note that it was not possible to develop the repo for spack without it, which is why I've included them together. I'll try to make a new branch that has only the docker changes, and then come back to this one after.

**Comment by vsoch** [comment]
See https://github.com/repology/repology-updater/pull/1174. Thank you!

**Comment by vsoch** [comment]
@AMDmi3 I've made the changes as requested. I'm not able to test anything without my development environment so I hope you can take it from here. Thank you!

**Comment by AMDmi3** [comment]
> I'm not able to test anything without my development environment

You obviously don't need to push docker stuff to the main repo in order to use it to develop and test the code.

Anyway, the code is ready to go, but there are some problems with data:

- Some entries (`ibm-java`) have empty list as a `version`.
- Downloads are out of sync with the version (zsh as example, version is `5.8`, but downloads list `http://downloads.sourceforge.net/project/zsh/zsh/5.4.2/zsh-5.4.2.tar.xz`).
- Information on version's `branch` is required in the dump, so we could handle rolling versions properly.
- As I understand, only the first version for a package gets to the dump. This is suboptimal as in many cases this leads to Repology seeing a rolling branch and no actual releases. I suggest to list all versions in the dump - so Repology is able to pick a subset on its side (we don't handle too many versions from a single repo well yet, but we can start by picking topmost release or topmost N releases for a package). Preferred format would be to group versions under a single package entry, so single pass parsing is possible and data common to the package is not duplicated.

It would also be nice to add information on patches (just list of file names) right away.

PS. As I can see, patches and depends are tied to the versions (as in `patch('cmake-macos-add-coreservices.patch', when='@3.11.0:3.13.3'`), note that this should be respected when listing depends and patches for a version.

**Review by AMDmi3 — APPROVED** [review]

**Comment by vsoch** [comment]
hey @AMDmi3 I can take a shot at fixing some of the data, and here are quick answers to your questions:

> Some entries (ibm-java) have empty list as a version.

This was a missed parsing on my part - [ibm-java](https://github.com/spack/spack/blob/24c01d57cfb99946eff3af7ec62b71ae6ec72ed8/var/spack/repos/builtin/packages/ibm-java/package.py#L25) uses a version list that doesn't seem to parse into traditional spack versions. I've updated the script now so this is fixed (and looks for both)

> Downloads are out of sync with the version (zsh as example, version is 5.8, but downloads list http://downloads.sourceforge.net/project/zsh/zsh/5.4.2/zsh-5.4.2.tar.xz).

Ah, that looks like another bug (and related to how spack packages are designed). If you look here https://github.com/spack/spack/blob/24c01d57cfb99946eff3af7ec62b71ae6ec72ed8/var/spack/repos/builtin/packages/zsh/package.py you'll see that the download defaults to 5.4, and there is actually a function that is used to map a version into that string (which I didn't use). I'll see if I can fix that so the most recent version matches the download.

> Information on version's branch is required in the dump, so we could handle rolling versions properly.

This is always going to be a nightly update from the develop branch on spack. The packages themselves come from all over the place so you couldn't designate a branch. I'll add a field "spack_branch" to each entry that will always be "develop."

> As I understand, only the first version for a package gets to the dump. This is suboptimal as in many cases this leads to Repology seeing a rolling branch and no actual releases. I suggest to list all versions in the dump - so Repology is able to pick a subset on its side (we don't handle too many versions from a single repo well yet, but we can start by picking topmost release or topmost N releases for a package). Preferred format would be to group versions under a single package entry, so single pass parsing is possible and data common to the package is not duplicated.

Oh I didn't realize I could provide them all! Yes of course I'll add this.

> It would also be nice to add information on patches (just list of file names) right away.

I'll add these too, and with the "when"

okay - data is updated! Please take a look!

https://raw.githubusercontent.com/spack/packages/main/data/repology.json

File: repology/parsers/parsers/spack.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                pkg.set_version(pkgdata['version'], normalize_version)
+                pkg.set_rawversion(pkgdata['version'])
+
+                # We don't currently have good keywords, could in the future!
+                # pkg.add_categories(packagedata['keywords'])
+                pkg.add_homepages(pkgdata['homepages'])
+                pkg.add_downloads(pkgdata['downloads'])
+                pkg.set_summary(pkgdata['summary'])
+                # Ensure that maintainer usernames are in context of spack
+                pkg.add_maintainers(["%s @spack" %x for x in pkgdata['maintainers']])
```
Comment (vsoch) [inline repology/parsers/parsers/spack.py:44]: Would we just add `pkg.add_patches(pkgdata['patches'])` here for newly added patches?

**Comment by AMDmi3** [comment]
> I'll add a field "spack_branch" to each entry that will always be "develop."

That's not what I've meant. I've meant a branch for package versions like this one:

https://github.com/spack/spack/blob/d00fc55e419a6ed18546d4a8b3fe3f39de641abc/var/spack/repos/builtin/packages/fides/package.py#L17

> I'll add these too, and with the "when"
>
> Please take a look!

Not quite. It doesn't allow to match downloads to versions for instance, and I cannot parse `when`'s myself.
What is needed instead of:

```json
        "papi": {
            "name": "papi", 
            "version": [
                "master",
                "6.0.0.1",
                "6.0.0",
		...
            ],
            "summary": "...",
            "maintainers": [],
            "licenses": {},
            "downloads": [
                "http://icl.cs.utk.edu/projects/papi/downloads/papi-master.tar.gz",
                "http://icl.cs.utk.edu/projects/papi/downloads/papi-6.0.0.1.tar.gz",
                "http://icl.cs.utk.edu/projects/papi/downloads/papi-6.0.0.tar.gz",
		...
            ],
            "homepages": [
                "http://icl.cs.utk.edu/papi/index.html"
            ],
            "patches": [
                "https://bitbucket.org/icl/papi/commits/53de184a162b8a7edff48fed01a15980664e15b1/raw when @5.4.0:5.6.99%gcc@8:",
                "crayftn-fixes.patch when @6.0.0:%cce@9:"
            ],
            "dependencies": [
                "lm-sensors",
                "cuda"
            ],
        },
```

is more like

```json
        "papi": {
            "name": "papi", 
            "versions": [
	        {
		    "version": "master",
		    "branch": "master",
		    "downloads": [
                        "http://icl.cs.utk.edu/projects/papi/downloads/papi-master.tar.gz",
		    ],
                    "patches": [
		    ],
                    "dependencies": [
                        "lm-sensors",
                        "cuda"
                    ]
		},
		{
		    "version": "6.0.0.1",
		    "downloads": [
                        "http://icl.cs.utk.edu/projects/papi/downloads/papi-6.0.0.1.tar.gz",
		    ],
                    "patches": [
		    ],
                    "dependencies": [
                        "lm-sensors",
                        "cuda"
                    ]
		},
		{
		    "version": "6.0.0",
		    "downloads": [
                        "http://icl.cs.utk.edu/projects/papi/downloads/papi-6.0.0.tar.gz",
		    ],
                    "patches": [
                        "crayftn-fixes.patch"
                    ],
                    "dependencies": [
                        "lm-sensors",
                        "cuda"
                    ]
		},
		...
            ],
            "summary": "...",
            "maintainers": [...],
            "licenses": {},
            "homepages": [
                "http://icl.cs.utk.edu/papi/index.html"
            ]
        },
```

**Comment by vsoch** [comment]
Thank you for showing me what you'd like instead of having me guess, my mind reading is a little rusty :)

I can update to get deps for each version, but that's going to increase the time to run the script by easily 1000 times - we would need to concretize 5000 times X the number of versions. I'm not sure is this is feasible to run for GitHub actions, which only has 6 hours.

**Comment by vsoch** [comment]
Okay, I've updated the script to include code for how to get all deps for each version, but it's commented out because it's just not feasible to run. I've made a best effort to get the data into the format you've requested:

https://raw.githubusercontent.com/spack/packages/main/data/repology.json

Note that dependencies are still on the level of the package, not the version.

Repology can be a good avenue into discovering packages in spack, but I don't see it being feasible to do all those concretizations to essentially reproduce what spack offers. If a user wants a package in spack, they can discover it along with the version, be able to see the (global) dependencies required by the package, and install it with spack if they desire.

**Comment by AMDmi3** [comment]
> script by easily 1000 times - we would need to concretize 5000 times X the number of versions

Not sure where 1000 times comes from, it looks like there are just ~20k package versions in total, which is just 4x of ~5k packages.

It looks almost good now, in fact I can live fine with generalized dependencies, just the patch info is gone now - if there's problem with per-version list of patches I probably can live with generalized list of patches too. Regardless of that, I'm going to deploy spack support tomorrow.

**Comment by vsoch** [comment]
The issue is the time - it's pretty slow to concretize, so yes 20K concretizations is going to take hours. You can look at the generate script and uncomment the known bits, and run with spack python to see for yourself (instructions in the README).

https://github.com/spack/packages/#usage

The issue with patches is really the same as deps - you'd need to concretize to get the explicit details because the "when" is provided as a conditional string that needs to be parsed. So our best bet is to add patches back to the top level with those strings, which I've done. Cheers!

**Comment by AMDmi3** [comment]
Another nit: https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/dbow2/package.py#L16 for this one `branch` is missing in the dump.

**Comment by vsoch** [comment]
The only way I could distinguish a branch is to use `isdevelop()`, which checks for a hard coded list of names like master, main, etc. If this is not going to work you can ask spack maintainers to take a look, they might have better ideas, and point them at the packages repo to update the metadata file. Thanks!

**Comment by AMDmi3** [comment]
Spack was added

**Pull request closed without merging**